Tasks/JavaToolInstallerV0/javatoolinstaller.ts

```diff
@@ -60,44 +52,161 @@ async function getJava(versionSpec: string) {
+
+        await taskutils.attach(sourceFile);
+    
+        const volumePath: string = getVolumePath(volumes);
+
+        const pkgPath: string = getPackagePath(volumePath);
+        try {
+            jdkDirectory = await installPkg(pkgPath, extendedJavaHome, versionSpec);
+        } finally {
+            // In case of an error, there is still a need to detach the disk image
+            await taskutils.detach(volumePath);
+        }
+    }
+    else if (fileExtension === '.pkg' && os.platform() === 'darwin') {
+        jdkDirectory = await installPkg(sourceFile, extendedJavaHome, versionSpec);
+    }
+    else {
+        // unpack the archive, set `JAVA_HOME` and save it for further processing
+        const extractDirectoryName: string = `${extendedJavaHome}_${JavaFilesExtractor.getStrippedName(sourceFile)}_${fileExtension.substr(1)}`;
+        const extractionDirectory: string = path.join(archiveExtractLocation, extractDirectoryName);
+        await unpackArchive(extractionDirectory, sourceFile, fileExtension, cleanDestinationDirectory);
+        jdkDirectory = JavaFilesExtractor.setJavaHome(extractionDirectory);
+    }
+    return jdkDirectory;
+}
+
+async function unpackArchive(unpackDir: string, jdkFileName: string, fileExt: string, cleanDestinationDirectory: boolean): Promise<void> {
+    const javaFilesExtractor = new JavaFilesExtractor();
+    if (!cleanDestinationDirectory && taskLib.exist(unpackDir)){
+        // do nothing since the files were extracted and ready for using
+        console.log(taskLib.loc('ArchiveWasExtractedEarlier'));
+    } else {
+        // unpack files to specified directory
+        console.log(taskLib.loc('ExtractingArchiveToPath', unpackDir));
+        await javaFilesExtractor.unzipJavaDownload(jdkFileName, fileExt, unpackDir);
+    }
+}
+
+/**
+ * Get the path to a folder inside the VOLUMES_FOLDER.
+ * Only for macOS.
+ * @param volumes VOLUMES_FOLDER contents before attaching a disk image.
+ * @returns string
+ */
+function getVolumePath(volumes: Set<string>): string {
+    const newVolumes: string[] = fs.readdirSync(VOLUMES_FOLDER).filter(volume => !volumes.has(volume));
+
+    if (newVolumes.length !== 1) {
+        throw new Error(taskLib.loc('UnsupportedDMGStructure'));
+    }
+    return path.join(VOLUMES_FOLDER, newVolumes[0]);
+}
+
+/**
+ * Get path to a .pkg file.
+ * Only for macOS.
+ * @param volumePath Path to the folder containing a .pkg file.
+ * @returns string
+ */
+function getPackagePath(volumePath: string): string {
+    const packages: string[] = fs.readdirSync(volumePath).filter(file => file.endsWith('.pkg'));
+
+    if (packages.length === 1) {
+        return path.join(volumePath, packages[0]);
+    } else if (packages.length === 0) {
+        throw new Error(taskLib.loc('NoPKGFile'));
+    } else {
+        throw new Error(taskLib.loc('SeveralPKGFiles'));
+    }
+}
+
+async function installPkg(pkgPath: string, extendedJavaHome: string, versionSpec: string): Promise<string> {
```

Comment:
Could you add js doc to these methods also?

Reply:
Sure, added